### ISSUE TITLE
How To Fix Laratrust Attach and Sync with Multiple Table Problem ?

### ISSUE THREAD

**Issue by lypsisrudiansyah** [santigarcor/laratrust#381]
i have used laratrust roles which are used on multiple tables (2 tables, for example, User and Cross) for auth

I have problems when attaching and syncing

every time I do an attachRole / syncRoles action for example I do in the User Model with id 1 then the Cross Model with id 1 is also attached or synchronized by the same role.

i have used laratrust 5.2
and laravel 6.6.0 version

I have set in config/laratrust.php

'user_models' => [
        'users' => 'App \ User',
        'crosses' => 'App \ Cross'
  ],

#ThanksInAdvance

**Comment by DJPretorius** [comment]
I am having a similar problem when calling `->roles` or `->hasRole('some_role_name')` on models with the same id.

```
'user_models' => [
    'modelAs' => 'App\ModelA',
    'modelBs' => 'App\ModelB'
],
```

I am using Redis for caching and I noticed I get the correct results when turning `cache.enabled` to false in the `config/laratrust.php` file.
BUT I really want to keep using caching since we are checking roles quite a lot.
Does anyone have some advice on this?

**Comment by santigarcor** [comment]
@DJPretorius what version are you using? have you overwritten the roles relationship inside the models?

**Comment by DJPretorius** [comment]
Hi @santigarcor 
I am using version `5.2.9`

No I didn't overwrite any laratrust functions. I only added the `LaratrustUserTrait` to each model.
The documentation for using multiple user models says:

> The value of the key in the key => value pair defines the name of the relationship inside the Role and Permission models.

And calling `$role->modelAs` or `$role->modelBs` works.

**Comment by santigarcor** [comment]
@DJPretorius can you please share with me what are you doing so I can try to replicate the issue?

**Comment by DJPretorius** [comment]
My Laratrust config.
```
return [
    'use_morph_map' => false,
    'checker' => 'default',
    'cache' => [
        'enabled' => true,     // <-- I am using Redis as our cache
        'expiration_time' => 3600,
    ],
    'use_teams' => false,
    'teams_strict_check' => false,
    'user_models' => [         
        'users' => 'App\User',
        'organisations' => 'App\Organisation'.   // <-- I added a second user model.
    ],
    'models' => [
        'role' => 'App\Role',
        'permission' => 'App\Permission',
        'team' => 'App\Team',
    ],
    'tables' => [
        'roles' => 'roles',
        'permissions' => 'permissions',
        'teams' => 'teams',
        'role_user' => 'role_user',
        'permission_user' => 'permission_user',
        'permission_role' => 'permission_role',
    ],
    'foreign_keys' => [
        'user' => 'user_id',
        'role' => 'role_id',
        'permission' => 'permission_id',
        'team' => 'team_id',
    ],
];
```

Then I added the `LaratrustUserTrait` to both models and then linked the models to roles via `role_users` like this:
```
|user_id    |role_id    |user_type        |
|-----------|-----------|-----------------|
|3          |6          |App\User         |
|3          |7          |App\User         |
|3          |8          |App\Organisation |
```

Now when I call `Organisation::find(3)->hasRole('the_name_of_role_with_id_8')` it returns `false`.
And when I call `Organisation::find(3)->roles` it returns `[roles_that_are_linked_to_user_with_id_3]`
So the problem is that the two user types are being confused and I suspect it is due to caching because the key saved in Redis is `laratrust_roles_for_user_3` witch does not differentiate between the two user types.

If you want to look at another explanation of my problem then here is [my post on Stack Overflow](https://stackoverflow.com/q/64119688/4232765)

**Comment by DJPretorius** [comment]
My Main concern is with the `->hasRole` function because I use it quite a lot. I really want it to use caching so that I don't hit the database every time I call `->hasRole`. 
Just for interest sake, have you had this use case before and is it supposed to store two different cache entries?

**Comment by santigarcor** [comment]
@DJPretorius I'll try to replicate that issue and I'll let you know.

Regarding the cache, I believe I already use caching for roles in that method

**Comment by DJPretorius** [comment]
Thank you. I appreciate the effort. 
Yes, I know you use caching in the `->hasRole` method. My question was "is it supposed to store two different cache entries?". Meaning is it supposed to store one cache entry for `user with id 3` and one entry for `organisation with id 3`?

**Comment by santigarcor** [comment]
@DJPretorius you're right, in theory it should store two different keys

**Comment by DJPretorius** [comment]
Hi @santigarcor. I took a deeper look at how you store values in the cache.
In the `LaratrustUserDefaultChecker` the cache key is built up using `$cacheKey = 'laratrust_roles_for_'.$this->userModelCacheKey() .'_'. $this->user->getKey();` and the `userModelCacheKey()` function is as follows:
```
public function userModelCacheKey()
{
    if (!Config::get('laratrust.use_morph_map')) {
        return 'user';
    }

    foreach (Config::get('laratrust.user_models') as $key => $model) {
        if ($this->user instanceof $model) {
            return $key;
        }
    }
}
```

At the moment I am not using the `morph_map`, thus the key will be the same for models with the same ID. This is the cause of the confusion.
A quick test showed that if I set `'use_morph_map' => true` in the laratrust config and change the `user_type` column to match the `key` of `user_models` then there are two cache entries corresponding to my two models. And now `->hasRole` returns the correct result!

Note: I had to clear my cache before the test because it would show the wrong stored values.

PS: I do not believe that the `use_morph_map` setting should change the key used by the cache. The cache's key should always contain the `key` part of `user_models`

Kind regards.
